File: PolicyPlus/FilterOptions.vb
Public Class FilterOptions
    Public CurrentFilter As FilterConfiguration
    Dim ProductNodes As Dictionary(Of PolicyPlusProduct, TreeNode)
    Public Function PresentDialog(Configuration As FilterConfiguration, Workspace As AdmxBundle) As DialogResult
        CurrentFilter = Nothing
        ProductNodes = New Dictionary(Of PolicyPlusProduct, TreeNode)
        AllowedProductsTreeview.Nodes.Clear()
        Dim addProductsToNodeCollection As Action(Of IEnumerable(Of PolicyPlusProduct), TreeNodeCollection)
        addProductsToNodeCollection = Sub(Products As IEnumerable(Of PolicyPlusProduct), Nodes As TreeNodeCollection)
                                          For Each product In Products
                                              Dim node = Nodes.Add(product.DisplayName)
                                              ProductNodes.Add(product, node)
                                              node.Tag = product
                                              If product.Children IsNot Nothing Then addProductsToNodeCollection(product.Children, node.Nodes)
                                          Next
                                      End Sub
        addProductsToNodeCollection(Workspace.Products.Values, AllowedProductsTreeview.Nodes) ' Recursively add products
        PrepareDialog(Configuration)
        Return ShowDialog()
    End Function
    Sub PrepareDialog(Configuration As FilterConfiguration)
        ' Set the UI element state from the current filter
        If Configuration.ManagedPolicy.HasValue Then
            PolicyTypeCombobox.SelectedIndex = If(Configuration.ManagedPolicy.Value, 1, 2)
        Else
            PolicyTypeCombobox.SelectedIndex = 0
        End If
        If Configuration.PolicyState.HasValue Then
            Select Case Configuration.PolicyState.Value
                Case FilterPolicyState.NotConfigured
                    PolicyStateCombobox.SelectedIndex = 1
                Case FilterPolicyState.Configured
                    PolicyStateCombobox.SelectedIndex = 2
                Case FilterPolicyState.Enabled
                    PolicyStateCombobox.SelectedIndex = 3
                Case FilterPolicyState.Disabled
                    PolicyStateCombobox.SelectedIndex = 4
            End Select
        Else
            PolicyStateCombobox.SelectedIndex = 0
        End If
        If Configuration.Commented.HasValue Then
            CommentedCombobox.SelectedIndex = If(Configuration.Commented.Value, 1, 2)
        Else
            CommentedCombobox.SelectedIndex = 0
        End If
        For Each node In ProductNodes.Values
            node.Checked = False
        Next
        If Configuration.AllowedProducts Is Nothing Then
            SupportedCheckbox.Checked = False
            AlwaysMatchAnyCheckbox.Checked = True
            MatchBlankSupportCheckbox.Checked = True
        Else
            SupportedCheckbox.Checked = True
            For Each product In Configuration.AllowedProducts
                ProductNodes(product).Checked = True
            Next
            AlwaysMatchAnyCheckbox.Checked = Configuration.AlwaysMatchAny
            MatchBlankSupportCheckbox.Checked = Configuration.MatchBlankSupport
            ' Expand to show all products with a different check state than their parent
            Dim expandIfNecessary As Action(Of TreeNode)
            expandIfNecessary = Sub(Node As TreeNode)
                                    For Each subnode As TreeNode In Node.Nodes
                                        expandIfNecessary(subnode)
                                        If subnode.IsExpanded Or subnode.Checked <> Node.Checked Then
                                            Node.Expand()
                                        End If
                                    Next
                                End Sub
            For Each node As TreeNode In AllowedProductsTreeview.Nodes
                expandIfNecessary(node)
            Next
        End If
    End Sub
    Private Sub SupportedCheckbox_CheckedChanged(sender As Object, e As EventArgs) Handles SupportedCheckbox.CheckedChanged
        RequirementsBox.Enabled = SupportedCheckbox.Checked
    End Sub
    Private Sub FilterOptions_KeyDown(sender As Object, e As KeyEventArgs) Handles Me.KeyDown
        If e.KeyCode = Keys.Escape Then DialogResult = DialogResult.Cancel
    End Sub
    Private Sub ResetButton_Click(sender As Object, e As EventArgs) Handles ResetButton.Click
        PrepareDialog(New FilterConfiguration)
    End Sub
    Private Sub OkButton_Click(sender As Object, e As EventArgs) Handles OkButton.Click
        ' Create a filter configuration object from the user's settings
        Dim newConf As New FilterConfiguration
        Select Case PolicyTypeCombobox.SelectedIndex
            Case 1
                newConf.ManagedPolicy = True
            Case 2
                newConf.ManagedPolicy = False
        End Select
        Select Case PolicyStateCombobox.SelectedIndex
            Case 1
                newConf.PolicyState = FilterPolicyState.NotConfigured
            Case 2
                newConf.PolicyState = FilterPolicyState.Configured
            Case 3
                newConf.PolicyState = FilterPolicyState.Enabled
            Case 4
                newConf.PolicyState = FilterPolicyState.Disabled
        End Select
        Select Case CommentedCombobox.SelectedIndex
            Case 1
                newConf.Commented = True
            Case 2
                newConf.Commented = False
        End Select
        If SupportedCheckbox.Checked Then
            newConf.AlwaysMatchAny = AlwaysMatchAnyCheckbox.Checked
            newConf.MatchBlankSupport = MatchBlankSupportCheckbox.Checked
            newConf.AllowedProducts = New List(Of PolicyPlusProduct)
            For Each kv In ProductNodes
                If kv.Value.Checked Then newConf.AllowedProducts.Add(kv.Key)
            Next
        End If
        CurrentFilter = newConf
        DialogResult = DialogResult.OK
    End Sub
    Private Sub AllowedProductsTreeview_EnabledChanged(sender As Object, e As EventArgs) Handles AllowedProductsTreeview.EnabledChanged
        ' Without this, the tree view looks really bad when disabled
        AllowedProductsTreeview.BackColor = If(AllowedProductsTreeview.Enabled, SystemColors.Window, SystemColors.Control)
    End Sub
    Private Sub AllowedProductsTreeview_AfterCheck(sender As Object, e As TreeViewEventArgs) Handles AllowedProductsTreeview.AfterCheck
        ' Recursively set the check state on child products only in response to the user
        If e.Action = TreeViewAction.Unknown Then Return
        PropogateCheckStateDown(e.Node)
    End Sub
    Sub PropogateCheckStateDown(Node As TreeNode)
        For Each subnode As TreeNode In Node.Nodes
            subnode.Checked = Node.Checked
            PropogateCheckStateDown(subnode)
        Next
    End Sub
End Class
Public Enum FilterPolicyState
    Configured
    NotConfigured
    Enabled
    Disabled
End Enum
Public Class FilterConfiguration
    Public ManagedPolicy As Boolean?
    Public PolicyState As FilterPolicyState?
    Public Commented As Boolean?
    Public AllowedProducts As List(Of PolicyPlusProduct)
    Public AlwaysMatchAny As Boolean
    Public MatchBlankSupport As Boolean
End Class
' The TreeView control has a bug: the displayed check state gets out of sync with the Checked property when the checkbox is double-clicked
' Fix adapted from https://stackoverflow.com/a/3174824
Friend Class DoubleClickIgnoringTreeView
    Inherits TreeView
    Protected Overrides Sub WndProc(ByRef m As Message)
        ' Ignore WM_LBUTTONDBLCLK
        If m.Msg <> &H203 Then MyBase.WndProc(m)
    End Sub
End Class

File: PolicyPlus/FilterOptions.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Windows.Forms;

namespace PolicyPlus
{
    public partial class FilterOptions
    {
        public FilterConfiguration CurrentFilter;
        private Dictionary<PolicyPlusProduct, TreeNode> ProductNodes;

        public FilterOptions()
        {
            InitializeComponent();
        }
        public DialogResult PresentDialog(FilterConfiguration Configuration, AdmxBundle Workspace)
        {
            CurrentFilter = null;
            ProductNodes = new Dictionary<PolicyPlusProduct, TreeNode>();
            AllowedProductsTreeview.Nodes.Clear();
            Action<IEnumerable<PolicyPlusProduct>, TreeNodeCollection> addProductsToNodeCollection;
            addProductsToNodeCollection = new Action<IEnumerable<PolicyPlusProduct>, TreeNodeCollection>((Products, Nodes) => { foreach (var product in Products) { var node = Nodes.Add(product.DisplayName); ProductNodes.Add(product, node); node.Tag = product; if (product.Children is not null) addProductsToNodeCollection(product.Children, node.Nodes); } });
            addProductsToNodeCollection(Workspace.Products.Values, AllowedProductsTreeview.Nodes); // Recursively add products
            PrepareDialog(Configuration);
            return ShowDialog();
        }
        public void PrepareDialog(FilterConfiguration Configuration)
        {
            // Set the UI element state from the current filter
            if (Configuration.ManagedPolicy.HasValue)
            {
                PolicyTypeCombobox.SelectedIndex = Configuration.ManagedPolicy.Value ? 1 : 2;
            }
            else
            {
                PolicyTypeCombobox.SelectedIndex = 0;
            }
            if (Configuration.PolicyState.HasValue)
            {
                switch (Configuration.PolicyState.Value)
                {
                    case FilterPolicyState.NotConfigured:
                        {
                            PolicyStateCombobox.SelectedIndex = 1;
                            break;
                        }
                    case FilterPolicyState.Configured:
                        {
                            PolicyStateCombobox.SelectedIndex = 2;
                            break;
                        }
                    case FilterPolicyState.Enabled:
                        {
                            PolicyStateCombobox.SelectedIndex = 3;
                            break;
                        }
                    case FilterPolicyState.Disabled:
                        {
                            PolicyStateCombobox.SelectedIndex = 4;
                            break;
                        }
                }
            }
            else
            {
                PolicyStateCombobox.SelectedIndex = 0;
            }
            if (Configuration.Commented.HasValue)
            {
                CommentedCombobox.SelectedIndex = Configuration.Commented.Value ? 1 : 2;
            }
            else
            {
                CommentedCombobox.SelectedIndex = 0;
            }
            foreach (var node in ProductNodes.Values)
                node.Checked = false;
            if (Configuration.AllowedProducts is null)
            {
                SupportedCheckbox.Checked = false;
                AlwaysMatchAnyCheckbox.Checked = true;
                MatchBlankSupportCheckbox.Checked = true;
            }
            else
            {
                SupportedCheckbox.Checked = true;
                foreach (var product in Configuration.AllowedProducts)
                    ProductNodes[product].Checked = true;
                AlwaysMatchAnyCheckbox.Checked = Configuration.AlwaysMatchAny;
                MatchBlankSupportCheckbox.Checked = Configuration.MatchBlankSupport;
                // Expand to show all products with a different check state than their parent
                Action<TreeNode> expandIfNecessary;
                expandIfNecessary = new Action<TreeNode>((Node) => { foreach (TreeNode subnode in Node.Nodes) { expandIfNecessary(subnode); if (subnode.IsExpanded | subnode.Checked != Node.Checked) { Node.Expand(); } } });
                foreach (TreeNode node in AllowedProductsTreeview.Nodes)
                    expandIfNecessary(node);
            }
        }
        private void SupportedCheckbox_CheckedChanged(object sender, EventArgs e)
        {
            RequirementsBox.Enabled = SupportedCheckbox.Checked;
        }
        private void FilterOptions_KeyDown(object sender, KeyEventArgs e)
        {
            if (e.KeyCode == Keys.Escape)
                DialogResult = DialogResult.Cancel;
        }
        private void ResetButton_Click(object sender, EventArgs e)
        {
            PrepareDialog(new FilterConfiguration());
        }
        private void OkButton_Click(object sender, EventArgs e)
        {
            // Create a filter configuration object from the user's settings
            var newConf = new FilterConfiguration();
            switch (PolicyTypeCombobox.SelectedIndex)
            {
                case 1:
                    {
                        newConf.ManagedPolicy = true;
                        break;
                    }
                case 2:
                    {
                        newConf.ManagedPolicy = false;
                        break;
                    }
            }
            switch (PolicyStateCombobox.SelectedIndex)
            {
                case 1:
                    {
                        newConf.PolicyState = FilterPolicyState.NotConfigured;
                        break;
                    }
                case 2:
                    {
                        newConf.PolicyState = FilterPolicyState.Configured;
                        break;
                    }
                case 3:
                    {
                        newConf.PolicyState = FilterPolicyState.Enabled;
                        break;
                    }
                case 4:
                    {
                        newConf.PolicyState = FilterPolicyState.Disabled;
                        break;
                    }
            }
            switch (CommentedCombobox.SelectedIndex)
            {
                case 1:
                    {
                        newConf.Commented = true;
                        break;
                    }
                case 2:
                    {
                        newConf.Commented = false;
                        break;
                    }
            }
            if (SupportedCheckbox.Checked)
            {
                newConf.AlwaysMatchAny = AlwaysMatchAnyCheckbox.Checked;
                newConf.MatchBlankSupport = MatchBlankSupportCheckbox.Checked;
                newConf.AllowedProducts = new List<PolicyPlusProduct>();
                foreach (var kv in ProductNodes)
                {
                    if (kv.Value.Checked)
                        newConf.AllowedProducts.Add(kv.Key);
                }
            }
            CurrentFilter = newConf;
            DialogResult = DialogResult.OK;
        }
        private void AllowedProductsTreeview_EnabledChanged(object sender, EventArgs e)
        {
            // Without this, the tree view looks really bad when disabled
            AllowedProductsTreeview.BackColor = AllowedProductsTreeview.Enabled ? SystemColors.Window : SystemColors.Control;
        }
        private void AllowedProductsTreeview_AfterCheck(object sender, TreeViewEventArgs e)
        {
            // Recursively set the check state on child products only in response to the user
            if (e.Action == TreeViewAction.Unknown)
                return;
            PropogateCheckStateDown(e.Node);
        }
        public void PropogateCheckStateDown(TreeNode Node)
        {
            foreach (TreeNode subnode in Node.Nodes)
            {
                subnode.Checked = Node.Checked;
                PropogateCheckStateDown(subnode);
            }
        }
    }

    public enum FilterPolicyState
    {
        Configured,
        NotConfigured,
        Enabled,
        Disabled
    }

    public class FilterConfiguration
    {
        public bool? ManagedPolicy;
        public FilterPolicyState? PolicyState;
        public bool? Commented;
        public List<PolicyPlusProduct> AllowedProducts;
        public bool AlwaysMatchAny;
        public bool MatchBlankSupport;
    }
    // The TreeView control has a bug: the displayed check state gets out of sync with the Checked property when the checkbox is double-clicked
    // Fix adapted from https://stackoverflow.com/a/3174824
    internal class DoubleClickIgnoringTreeView : TreeView
    {
        protected override void WndProc(ref Message m)
        {
            // Ignore WM_LBUTTONDBLCLK
            if (m.Msg != 0x203)
                base.WndProc(ref m);
        }
    }
}